### PR TITLE
MM-14559 Rewrite existing plugin API tests to exclusively test via actual plugin

### PR DESCRIPTION
**Pull request by reflog**
#### Summary
Move from testing using internal plugin APIs to testing via RPC. Previously embedded Go code is moved to separate files.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/10438

### CODE REVIEW

**Comment by grundleborg** [comment]
@crspeller @levb would be great if you can give some feedback on the approach taken here.

**Comment by crspeller** [comment]
@reflog Looks good! Certainly heading in the right direction.
What might be cool is automatically traversing the tests folder and automatically running all the tests inside. (you would have to somehow specify your assertions and options inside the file) Anything to ease test development would encourage more tests to be written.

**Comment by reflog** [comment]
Hi @crspeller , I've went ahead with your suggestion and implemented a 'bulk test' function that scans a folder of tests. Each plugin test is responsible for it's own assertions based on the data passed to it.

I've split all basic tests into separate files (I defined 'basic' as any test that doesn't require any preparations)

Let me know what you think!

**Comment by crspeller** [comment]
@reflog Looking good! 
The only comment is it's a bit inconvenient to not have the usual assertion libraries available in the plugins. Maybe a test specific hook that passes the test context? 0/5 no idea if that would even work.

**Comment by reflog** [comment]
@crspeller - I'm not sure that passing test context through a specific hook is a viable idea and would add complexity that I think might be better to avoid. This way, the plugins don't have any 'test specific' code and the assertions are done via regular 'if' statements.

What do you think? Is the ticket merge-able in it's current state, or you want to pursue the 'test context' idea?

**Comment by crspeller** [comment]
@reflog  I think it should be mergeable without that suggestion. It's better than it was.

File: app/plugin_api_test.go [hunk 1]
```diff
@@ -376,45 +359,10 @@ func TestPluginAPILoadPluginConfiguration(t *testing.T) {
...
-				return err
-			}
-
-			return nil
-		}
-
-		func (p *MyPlugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*model.Post, string) {
-			return nil, fmt.Sprintf("%v%v%v", p.configuration.MyStringSetting, p.configuration.MyIntSetting, p.configuration.MyBoolSetting)
-		}
+	pwd, _ := os.Getwd()
```
Comment (levb) [inline app/plugin_api_test.go]: (nit) 1/5 maybe use `fileutils.FindDir("tests")` like in https://github.com/mattermost/mattermost-server/blob/master/api4/file_test.go#L30

**Comment by levb** [comment]
@reflog @crspeller Were you guys talking about passing the test context via an RPC, and invoking it remotely? I'm very curious if it works, maybe just pass it as a `testing.TB` interface so that there's no data serialization involved, or do you specifically need a `testing.T` for the assertions? Not sure if there's a difference in the mechanics, really.

File: app/plugin_api_test.go (hunk 1 above)
Comment (reflog) [reply app/plugin_api_test.go]: Done!

**Comment by reflog** [comment]
@levb - yes, @crspeller suggested maybe passing along the testing context by adding another RPC interface to the plugins, in order to support assertion libraries.
I think it's a little bit of an overkill, since the way it's implemented right now - each plugin test is written exactly like a regular plugin would, without any hacks and does testing 'of itself'.

I can start investigating the passing of the testing context if you feel that it's worth it. If not - let's merge :)

**Pull request merged**